data/src/main/java/com/linkedin/data/NonBlockingDataParser.java

```diff
@@ -0,0 +1,184 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.data;
+
+import java.io.IOException;
+
+
+/**
+ * Data parser interface invoked by non blocking decoder.
+ *
+ * This interface contains methods that are invoked when parsing a Data object.
+ * Each method represents a different kind of event/read action
+ *
+ * Methods can throw IOException as a checked exception to indicate parsing error.
+ *
+ * @author amgupta1
+ */
+public interface NonBlockingDataParser
```

Comment:
The other package is also in the same module, so there wouldn't be a cyclic dependency. 
I'm suggesting that as this is closely associated with the non blocking decoder and all our non-blocking implementation is in the data.codec.entitystream package.

Reply:
com.linkedin.data.codec.entitystream is sub package of com.linkedin.data
so ideally, com.linkedin.data shouldn't be using references from com.linkedin.data.codec.entitystream
which would happen is I move this to entitystream as ByteString use DataParser

Reply:
There is no convention against reference sub-packages. ByteString already does that. Even classes in java (Eg java.util.Map) has similar references.
I think we can also create a separate package `com.linkedin.data.parser` 

This parse doesn't fit with the other classes in data package.